src/psyclone/tests/psyir/frontend/where_handler_test.py

```diff
@@ -0,0 +1,345 @@
+    loop = fake_parent.children[0]
+    assert isinstance(loop.ast, Fortran2003.Where_Construct)
+    assert isinstance(loop, Loop)
+    assert isinstance(loop.loop_body[0], Loop)
+    assert isinstance(loop.loop_body[0].loop_body[0], Loop)
+    # Check that we have an IF block within the innermost loop
+    ifblock = loop.loop_body[0].loop_body[0].loop_body[0]
+    assert isinstance(ifblock, IfBlock)
+    assert isinstance(ifblock.ast, Fortran2003.Where_Construct)
+    assert isinstance(ifblock.condition, BinaryOperation)
+    assert ifblock.condition.operator == BinaryOperation.Operator.NE
+    assert ("ArrayReference[name:'ptsu']\n"
+            "Reference[name:'widx1']\n" in str(ifblock.condition.children[0]))
+    assert "Literal[value:'0." in str(ifblock.condition.children[1])
+    # Check that this IF block has an else body
+    assert ifblock.else_body is not None
+    assert isinstance(ifblock.else_body[0], Assignment)
+    assert isinstance(ifblock.else_body[0].lhs, Array)
+    assert ifblock.else_body[0].lhs.name == "z1_st"
+
+
+def test_where_stmt_validity(parser):
+    ''' Check that the correct exceptions are raised when the parse tree
+    for a WHERE statement has an unexpected structure. '''
+    from psyclone.psyGen import InternalError
+    fake_parent = Schedule()
+    processor = Fparser2Reader()
+    reader = FortranStringReader(
+        "WHERE( at_i(:,:) > rn_amax_2d(:,:) )   "
+        "a_i(:,:,jl) = a_i(:,:,jl) * rn_amax_2d(:,:) / at_i(:,:)")
+    fparser2spec = Fortran2003.Where_Stmt(reader)
+    # Break the parse tree
+    fparser2spec.items = (fparser2spec.items[0], "a string")
+    with pytest.raises(InternalError) as err:
+        processor.process_nodes(fake_parent, [fparser2spec], None)
+    assert "items tuple to be an Assignment_Stmt but found" in str(err.value)
+    # Break it so that the tuple only contains one item
+    fparser2spec.items = (fparser2spec.items[0], )
+    with pytest.raises(InternalError) as err:
+        processor.process_nodes(fake_parent, [fparser2spec], None)
+    assert ("Where_Stmt to have exactly two entries in 'items' but found 1"
+            in str(err.value))
+
+
+def test_where_stmt(parser):
+    ''' Basic check that we handle a WHERE statement correctly. '''
+    from psyclone.psyGen import Loop
+    fake_parent = Schedule()
+    processor = Fparser2Reader()
+    reader = FortranStringReader(
+        "WHERE( at_i(:,:) > rn_amax_2d(:,:) )   "
+        "a_i(:,:,jl) = a_i(:,:,jl) * rn_amax_2d(:,:) / at_i(:,:)")
+    fparser2spec = Fortran2003.Where_Stmt(reader)
+    processor.process_nodes(fake_parent, [fparser2spec], None)
+    fake_parent.view()
+    assert len(fake_parent.children) == 1
+    assert isinstance(fake_parent[0], Loop)
+
+
+def test_where_ordering(parser):
+    ''' Check that the generated schedule has the correct ordering when
+    a WHERE construct is processed. '''
+    from psyclone.psyGen import Assignment, Loop, CodeBlock
```

Comment:
W:324, 4: Redefining name 'CodeBlock' from outer scope (line 44) (redefined-outer-name)
W:324, 4: Reimport 'CodeBlock' (imported line 44) (reimported)

Reply:
We're good now.